lib/smack/SmackRep.cpp

```diff
@@ -1012,13 +1017,47 @@ std::string SmackRep::code(llvm::CallInst& ci) {
 
   std::string s = fmt;
   for (unsigned i=1; i<ci.getNumOperands()-1; i++) {
-    const Expr* a = arg(f, i, ci.getOperand(i));
+    Value* a = ci.getOperand(i);
     std::string::size_type idx = s.find('@');
     assert(idx != std::string::npos && "inline code: too many arguments.");
 
+    llvm::Type* target_type = a->getType();
+    bool is_cast = false;
+    if (idx + 1 < s.length()) {
+      switch(s[idx+1]) {
+      case 'c': {
+	target_type = Type::getInt8Ty(a->getContext());
+	is_cast = true;
+	break;
+      }
+      case 'f': {
+	target_type = Type::getFloatTy(a->getContext());
+	is_cast = true;
+	break;
+      }
+      case 'h': {
+	target_type = Type::getInt16Ty(a->getContext());
+	is_cast = true;
+	break;
+      }
+      case 'i': {
+	target_type = Type::getInt32Ty(a->getContext());
+	is_cast = true;
+	break;
+      }
+      default:
```

Comment:
I think we should have some light-weight error checking here. I think your code would currently silently accept something like `@x`, is that right? Maybe you could add some error checking code here that would throw an assertion violation if a wrong character is used.

Reply:
This behavior is intended. `@x` is interpreted as `<arg-expr>x`.

Reply:
I can't find any instances of `@[a-zA-Z]` in the existing code, but maybe it's useful.
Right now we allow a weird situation where `@x` becomes `<arg-expr>x` but `@f` is just `<precast-arg-expr>`. Perhaps a consistent solution would be to check in `default` if the second character is in `[a-zA-Z]` and if it's not on the list from https://docs.python.org/2.7/library/struct.html#format-characters we report an error. If we want the literal character there, we can escape it with something like `\`, so `@\x` is `<arg-expr>x`.

Reply:
I think we don't want to make the design complicated even if it means a more complete design. For our purposes, the current characters are sufficient.

Reply:
Ok, I get it. I am ok that it stays as is.